File: algorithm/validate-binary-search-tree/validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    
    TreeNode* findNode(TreeNode* root, TreeNode* node) {
        TreeNode* tmp = root;
        while (tmp) {
            if (tmp->val == node->val)
                return tmp;
            else if (tmp->val < node->val)
                tmp = tmp->right;
            else if (tmp->val > node->val)
                tmp = tmp->left;
        }
        return tmp;
    }
public:
    bool isValidBST(TreeNode* root) {

        vector<TreeNode*> nodes;
        queue<TreeNode*> BFS;
        TreeNode *tmp;
        
        BFS.push(root);
        while(!BFS.empty() && BFS.front()) {
            tmp = BFS.front();
            BFS.pop();
            if (tmp->left) {
                BFS.push(tmp->left);
                nodes.push_back(tmp->left);
            }
            if (tmp->right) {
                BFS.push(tmp->right);
                nodes.push_back(tmp->right);
            }
        }
        
        for (auto n: nodes) {
            if (n != findNode(root, n))
                return false;
        }
        return true;
    }
};

File: algorithm/validate-binary-search-tree/validate-binary-search-tree.go
package validate_binary_search_tree

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isValidBST(root *TreeNode) bool {
	_, _, ok := isValidBSTTraversal(root)
	return ok
}

func isValidBSTTraversal(root *TreeNode) (int, int, bool) {
	var lMin, lMax, rMin, rMax int
	var lValid, rValid bool

	if root.Right == nil && root.Left == nil {
		return root.Val, root.Val, true
	}

	if root.Left != nil {
		lMin, lMax, lValid = isValidBSTTraversal(root.Left)
		if !lValid {
			return -1, -1, false
		}

		if lMax >= root.Val {
			return -1, -1, false
		}
	}

	if root.Right != nil {
		rMin, rMax, rValid = isValidBSTTraversal(root.Right)
		if !rValid {
			return -1, -1, false
		}
		if rMin <= root.Val {
			return -1, -1, false
		}
	}

	if lValid && !rValid {
		return lMin, getMax(lMax, root.Val), true
	}

	if rValid && !lValid {
		return getMin(rMin, root.Val), rMax, true
	}

	min, max := root.Val, root.Val

	min = getMin(min, lMin)
	max = getMax(max, rMax)

	return min, max, true
}

func getMin(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func getMax(a, b int) int {
	if a > b {
		return a
	}
	return b
}
